File: clients/scala/src/main/scala/ai_cup_22/model/Action.scala
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Action unit is currently performing
 *
 * @param finishTick Tick when the action will be finished
 * @param actionType Type of the action
 */
case class Action(finishTick: Int, actionType: ai_cup_22.model.ActionType) {
    /**
     * Write Action to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeInt(stream, finishTick)
        actionType.writeTo(stream)
    }

    /**
     * Get string representation of Action
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Action { ")
        stringBuilder.append("finishTick: ")
        stringBuilder.append(finishTick)
        stringBuilder.append(", ")
        stringBuilder.append("actionType: ")
        stringBuilder.append(actionType)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Action {
    /**
     * Read Action from input stream
     */
    def readFrom(stream: java.io.InputStream): Action = Action(
        StreamUtil.readInt(stream),
        ai_cup_22.model.ActionType.readFrom(stream)
    )
}

File: clients/swift/Sources/AiCup22/Model/Action.swift
/// Action unit is currently performing
public struct Action {
    /// Tick when the action will be finished
    let finishTick: Int32

    /// Type of the action
    let actionType: ActionType

    /// Read Action from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Action {
        var finishTick: Int32
        finishTick = stream.readInt32()
        var actionType: ActionType
        actionType = ActionType.readFrom(stream)
        return Action(finishTick: finishTick, actionType: actionType)
    }

    /// Write Action to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(finishTick)
        actionType.writeTo(stream)
    }
}
